### PR TITLE
aerc: add module

### PR DESCRIPTION
**Pull request by wiIIiambonner**
### Description

Added a module for the `aerc` email client.

### Checklist

- [ ] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [ ] Code tested through `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [x] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

**Comment by podocarp** [comment]
Just a small thing I noticed, you added yourself as maintainer but did not specify `meta.maintainers` in the module

**Comment by ratsclub** [comment]
I tried to use the default configuration files when the user doesn't set it's own configurations here but I could not proceed due to this error:

```
error: A definition for option `xdg.configFile.aerc/aerc.conf.source' is not of type `path'. Definition values:
       - In `/home/victor/Projects/home-manager/modules/programs/aerc.nix': "@aerc@/share/aerc/aerc.conf"
(use '--show-trace' to show detailed location information)
```

This is the relevant piece of code used before:

```nix
xdg.configFile."aerc/aerc.conf".source = if cfg.conf != { } then
    formatIni.generate "aerc.conf" cfg.conf
  else
    "${pkgs.aerc}/share/aerc/aerc.conf";
```

File: modules/programs/aerc.nix [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    mkKeyValue = k: v:
+      let value = (if isBool v then boolToString else toString) v;
+      in "${k} = ${value}";
+  };
+
+  configDir =
+    if pkgs.stdenv.isDarwin then
+      "Library/Application Support/aerc"
+    else
+      "${config.xdg.configHome}/aerc";
```
Comment (polykernel) [inline modules/programs/aerc.nix:18]:
```suggestion
  configDir =
    if pkgs.stdenv.isDarwin then
      "Library/Application Support"
    else
      config.xdg.configHome;
```

File: modules/programs/aerc.nix [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    meta.maintainers = [ maintainers.ratsclub ];
+
+    home.packages = [ cfg.package ];
+
+    xdg.configFile = {
```
Comment (polykernel) [inline modules/programs/aerc.nix:114]:
```suggestion
    home.file.${configDir} = {
```

File: tests/modules/programs/aerc/with-config-file.nix [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        ''
+      }
+
+      assertFileContent home-files/.config/aerc/templates/foo ${
+        builtins.toFile "foo" ''
+          bar
+        ''
+      }
+    '';
+  };
```
Comment (polykernel) [inline tests/modules/programs/aerc/with-config-file.nix:67]:
```suggestion
    nmt.script = let
      configDir = if pkgs.stdenv.isDarwin then
          "home-files/Library/Application Support"
        else
          "home-files/.config";
      in ''
        assertFileExists ${configDir}/aerc/aerc.conf
        assertFileExists ${configDir}/aerc/binds.conf
        assertFileExists ${configDir}/aerc/stylesets/default
        assertFileExists ${configDir}/aerc/templates/foo
  
        assertFileContent ${configDir}/aerc/aerc.conf ${
          builtins.toFile "aerc.conf" ''
            [ui]
            mouse-enabled=true
            sort=-r date
          ''
        }
  
        assertFileContent ${configDir}/aerc/binds.conf ${
          builtins.toFile "binds.conf" ''
            [terminal]
            $ex=<C-x>
            $noinherit=true
          ''
        }
  
        assertFileContent ${configDir}/aerc/stylesets/default ${
          builtins.toFile "default" ''
            *.default = true
            *.selected.reverse = toggle
          ''
        }
  
        assertFileContent ${configDir}/aerc/templates/foo ${
          builtins.toFile "foo" ''
            bar
          ''
        }
      '';
  };
```

**Comment by polykernel** [comment]
> I tried to use the default configuration files when the user doesn't set it's own configurations here but I could not proceed due to this error:
> 
> ```
> error: A definition for option `xdg.configFile.aerc/aerc.conf.source' is not of type `path'. Definition values:
>        - In `/home/victor/Projects/home-manager/modules/programs/aerc.nix': "@aerc@/share/aerc/aerc.conf"
> (use '--show-trace' to show detailed location information)
> ```
> 
> This is the relevant piece of code used before:
> 
> ```nix
> xdg.configFile."aerc/aerc.conf".source = if cfg.conf != { } then
>     formatIni.generate "aerc.conf" cfg.conf
>   else
>     "${pkgs.aerc}/share/aerc/aerc.conf";
> ```

I believe this is caused by the following line:
```nix
test.stubs.aerc = { };
```

Defining `tests.stubs.<package>` will override the package globally because it internally uses an overlay. You can avoid this by only supplying the stub package to the package option.
```nix
package = config.lib.test.mkStubPackage { };
```

<br>

I think the default should be sourced from `cfg.package` to not make the configuration depend on the nixpkgs version but I don't think it matters too much here given it is only a fallback.

**Comment by ratsclub** [comment]
Thanks for the review!

> I think the default should be sourced from `cfg.package` to not make the configuration depend on the nixpkgs version but I don't think it matters too much here given it is only a fallback.

This makes me think that I shouldn't provide this default file as a simple modification of the package would break it...

File: modules/programs/aerc.nix (hunk 1 above)
Comment (ratsclub) [reply modules/programs/aerc.nix]: Done.

File: modules/programs/aerc.nix (hunk 2 above)
Comment (ratsclub) [reply modules/programs/aerc.nix]: Done.

File: tests/modules/programs/aerc/with-config-file.nix (hunk 3 above)
Comment (ratsclub) [reply tests/modules/programs/aerc/with-config-file.nix]: Done.

**Comment by teto** [comment]
do you plan to support hte generation of mailboxes via the mail module ?

**Comment by ratsclub** [comment]
I can definitely try to do it so!

**Comment by ratsclub** [comment]
@teto would it be able to merge the module as is and then I provide the account setup in another PR? I guess that it will be a lot more involved as it might support imap, notmuch and maildir and make the PR bigger to review.

**Comment by ratsclub** [comment]
Unfortunately, the [config/config.go](https://git.sr.ht/~rjarry/aerc/tree/fb0e9e3e41cd263de139de0bf620cc69ee5d61ae/item/config/config.go#L853) file makes it impossible to store `accounts.conf` on the nix store...

**Comment by teto** [comment]
> @teto would we be able to merge the module as is and then I provide the account setup in another PR? I guess that it will be a lot more involved as it might support imap, notmuch and maildir and make the PR bigger to review.

I understand. Yeah no pb. Other people have already reviewed the PR so I'll let them merge.

> Unfortunately, the [config/config.go](https://git.sr.ht/~rjarry/aerc/tree/fb0e9e3e41cd263de139de0bf620cc69ee5d61ae/item/config/config.go#L853) file makes it impossible to store accounts.conf on the nix store...

aerc should support a way that doesn't embed passwords in its config. I dont know if it's availaible or not but an aerc user may be able to negotiate a solution upstream.

**Comment by ratsclub** [comment]
> aerc should support a way that doesn't embed passwords in its config. I dont know if it's availaible or not but an aerc user may be able to negotiate a solution upstream.

The only way I know of is reading the emails from `maildir` or `notmuch` and sending them through `msmtp`. All the other ways require an `accounts.conf` with passwords on it.

**Comment by polykernel** [comment]
> This makes me think that I shouldn't provide this default file as a simple modification of the package would break it...

I think in general it is a good idea to keep the configuration and the package separate as much as possible to allow better composition but sometimes it is inevitable. I am not an aerc user myself but I trust you to make the right decision.

**Comment by ratsclub** [comment]
Thanks for the trust!

I feel that we are ready to merge as of now. I don't intend to implement the account setting right now as this will involve an upstream effort. May we get this approved?

**Comment by ratsclub** [comment]
I sent a patch upstream about the accounts too: https://lists.sr.ht/~rjarry/aerc-devel/patches/31289

**Comment by ambroisie** [comment]
@wiIIiambonner there is a way to add a command for passwords, rather than writing the password itself in the configuration file (see [here](https://git.sr.ht/~rjarry/aerc/commit/b8208509f47301bcb03bc222f5e5765833478250)).

Is there a reason you closed this MR?

**Comment by teto** [comment]
the new MR I suppose https://github.com/nix-community/home-manager/pull/3070

**Pull request closed without merging**